### PR TITLE
Small assortment of various LFS changes

### LINKED ISSUE #537: options/posix: mkostemp should ignore O_WRONLY
Since we pass `O_RDWR`, these will OR together to give `O_ACCMODE` which is not what we want. Technically (according to the man page) callers shouldn't pass this flag, but musl at least seems to work around this in the way I describe.

We should also probably move `mkostemp` to `options/glibc` instead.

### PR DESCRIPTION
**Pull request by no92**
Fixes #537.

### CODE REVIEW

File: options/posix/include/bits/posix/timer_t.h [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#ifndef _MLIBC_TIMER_T_H
+#define _MLIBC_TIMER_T_H
+
+typedef void * timer_t;
```
Comment (avdgrinten) [inline options/posix/include/bits/posix/timer_t.h:4]: Why is this a `void *`? What will it point to if POSIX timers are implemented?

Comment (no92) [reply options/posix/include/bits/posix/timer_t.h:4]: On both glibc and musl, `timer_t` is `typedef`ed as a `void *`.

Comment (avdgrinten) [reply options/posix/include/bits/posix/timer_t.h:4]: What does it point to?

Comment (no92) [reply options/posix/include/bits/posix/timer_t.h:4]: On musl, it points to a `intptr_t`. I remember reading that on glibc, it also points to some integer type.

File: options/linux/include/syscall.h [hunk 2]
```diff
@@ -0,0 +1 @@
+#include <sys/syscall.h>
```
Comment (64) [inline options/linux/include/syscall.h:1]: Why is this needed? This should probably live in the sysdeps folder since it's specific to the Linux kernel.

**Comment by 64** [comment]
The ABI diff fails because sysinfo was (correctly) changed to have C linkage.

File: options/posix/generic/posix_stdlib.cpp [hunk 3]
```diff
@@ -487,3 +488,8 @@ char *canonicalize_file_name(const char *) {
 	__ensure(!"Not implemented");
 	__builtin_unreachable();
 }
+
+int clearenv(void) {
```
Comment (64) [inline options/posix/generic/posix_stdlib.cpp:492]: This should live with the other environment functions.

File: options/posix/generic/sys-mman-stubs.cpp [hunk 4]
```diff
@@ -23,9 +23,17 @@ int mlock(const void *, size_t) {
 	__builtin_unreachable();
 }
 
-int mlockall(int) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int mlockall(int flags) {
+	if(!mlibc::sys_mlockall) {
```
Comment (64) [inline options/posix/generic/sys-mman-stubs.cpp:27]: This should use the checking macro

File: options/posix/generic/unistd-stubs.cpp [hunk 5]
```diff
@@ -487,7 +487,16 @@ long pathconf(const char *, int name) {
 }
 
 int pause(void) {
-	__ensure(!"Not implemented");
+	if(!mlibc::sys_pause) {
+		MLIBC_MISSING_SYSDEP();
```
Comment (64) [inline options/posix/generic/unistd-stubs.cpp:491]: Same here.

File: options/linux/include/syscall.h (hunk 2 above)
Comment (no92) [reply options/linux/include/syscall.h:1]: This header is present on both glibc and musl, and Dennis ran into some program expecting this to exist. Both of the other libcs just include `<sys/syscall.h>` as well.

Comment (64) [reply options/linux/include/syscall.h:1]: Ok, but it should still probably go in the Linux sysdeps. I can conceivably imagine programs checking has_include for this header and making decisions based on that, so it's better for it to not be present unless it's implemented correctly.

File: options/posix/generic/posix_stdlib.cpp (hunk 3 above)
Comment (no92) [reply options/posix/generic/posix_stdlib.cpp:492]: `clearenv` should probably only be included in the posix option, and even there it's kinda doubtful - it hasn't been accepted into the standard as far as I can tell. So I'm only moving this up to be right below `secure_getenv`.

Comment (64) [reply options/posix/generic/posix_stdlib.cpp]: The other functions in environment.cpp are behind a posix ifdef already, so IMO it'd make more sense to go there.

File: options/linux/include/syscall.h (hunk 2 above)
Comment (ArsenArsen) [reply options/linux/include/syscall.h:1]: Isn't sys/syscall.h also installed as a part of option/linux?

Comment (64) [reply options/linux/include/syscall.h:1]: There's one there, but I don't think it's used and we ought to remove it.

Comment (ArsenArsen) [reply options/linux/include/syscall.h:1]: hm? what do you mean?

Comment (64) [reply options/linux/include/syscall.h:1]: Check options/linux/include/sys/syscall.h and sysdeps/linux/include/sys/syscall.h

**Review by 64 — APPROVED** [review]

File: options/posix/generic/posix_stdlib.cpp [hunk 6]
```diff
@@ -474,6 +475,11 @@ char *secure_getenv(const char *name) {
 		return getenv(name);
 }
 
+int clearenv(void) {
+	__ensure(!"Not implemented");
```
Comment (ArsenArsen) [inline options/posix/generic/posix_stdlib.cpp:479]: Why not?

File: options/posix/generic/sys-time-stubs.cpp [hunk 7]
```diff
@@ -68,3 +68,8 @@ int setitimer(int which, const struct itimerval *new_value, struct itimerval *ol
 	}
 	return 0;
 }
+
+int timer_create(clockid_t, struct sigevent *__restrict, timer_t *__restrict) {
+	__ensure(!"Not implemented");
+	__builtin_unreachable();
```
Comment (ArsenArsen) [inline options/posix/generic/sys-time-stubs.cpp:74]: Didn't we have an implementation of this in some other PR (and can we add one if we didn't?)

File: options/posix/generic/unistd-stubs.cpp [hunk 8]
```diff
@@ -487,7 +487,12 @@ long pathconf(const char *, int name) {
 }
 
 int pause(void) {
-	__ensure(!"Not implemented");
+	MLIBC_CHECK_OR_ENOSYS(mlibc::sys_pause, -1);
+	if(int e = mlibc::sys_pause(); e) {
```
Comment (ArsenArsen) [inline options/posix/generic/unistd-stubs.cpp:491]: Is it possible for `sc_error(e) != e` in case of error? If so, this return value is wrong

File: sysdeps/linux/generic/sysdeps.cpp [hunk 9]
```diff
@@ -893,6 +893,24 @@ int sys_ttyname(int fd, char *buf, size_t size) {
 	return 0;
 }
 
+int sys_pause() {
+#ifdef SYS_pause
+	auto ret = do_syscall(SYS_pause);
+#else
+	auto ret = do_syscall(SYS_ppoll, 0, 0, 0, 0);
+#endif
```
Comment (ArsenArsen) [inline sysdeps/linux/generic/sysdeps.cpp:901]: Why?

**Review by ArsenArsen — CHANGES_REQUESTED** [review]

File: sysdeps/linux/generic/sysdeps.cpp (hunk 9 above)
Comment (64) [reply sysdeps/linux/generic/sysdeps.cpp:901]: Presumably RISC-V does not have a pause syscall.

Comment (no92) [reply sysdeps/linux/generic/sysdeps.cpp:901]: Indeed, RISC-V does not have pause.

File: options/posix/generic/sys-time-stubs.cpp (hunk 7 above)
Comment (no92) [reply options/posix/generic/sys-time-stubs.cpp:74]: Not that I would know of or could find. Given that the implementation is non-trivial and I wouldn't know anything that uses it, I'd leave it for a future PR.

**Comment by ArsenArsen** [comment]
BTW, can you add tests for the new functions?

**Review by 64 — APPROVED** [review]

**Review by Dennisbonke — APPROVED** [review]
LGTM

**Review by Geertiebear — APPROVED** [review]
LGTM.
